src/natcap/invest/stormwater.py

```diff
@@ -0,0 +1,1332 @@
+    """Calculate array of volumes (retention or infiltration) from arrays 
+    of precipitation values and stormwater ratios. This is meant to be used
+    with raster_calculator.
+
+    Args:
+        ratio_array (numpy.ndarray): 2D array of stormwater ratios. Assuming
+            that its nodata value is the global NODATA.
+        precip_array (numpy.ndarray): 2D array of precipitation amounts 
+            in millimeters/year
+        precip_nodata (float): nodata value for the precipitation array
+        pixel_area (float): area of each pixel in m^2
+    
+    Returns:
+        2D numpy.ndarray of precipitation volumes in m^3/year
+    """
+    volume_array = numpy.full(ratio_array.shape, NODATA, dtype=float)
+    valid_mask = (
+        (ratio_array != NODATA) & 
+        (precip_array != precip_nodata))
+
+    # precipitation (mm/yr) * pixel area (m^2) * 
+    # 0.001 (m/mm) * ratio = volume (m^3/yr)
+    volume_array[valid_mask] = (
+        precip_array[valid_mask] *
+        ratio_array[valid_mask] *
+        pixel_area * 0.001)
+    return volume_array
+
+
+def avoided_pollutant_load_op(lulc_array, lulc_nodata, retention_volume_array, 
+        sorted_lucodes, emc_array):
+    """Calculate avoided pollutant loads from LULC codes retention volumes.
+    This is intented to be used with pygeoprocessing.raster_calculator.
+
+    Args:
+        lulc_array (numpy.ndarray): 2D array of LULC codes
+        lulc_nodata (int): nodata value for the LULC array
+        retention_volume_array (numpy.ndarray): 2D array of stormwater 
+            retention volumes, with the same shape as ``lulc_array``. It is
+            assumed that the retention volume nodata value is the global NODATA
+        sorted_lucodes (numpy.ndarray): 1D array of the LULC codes in order
+            from smallest to largest
+        emc_array (numpy.ndarray): 1D array of pollutant EMC values for each 
+            lucode. ``emc_array[i]`` is the EMC for the LULC class at 
+            ``sorted_lucodes[i]``.
+
+    Returns:
+        2D numpy.ndarray with the same shape as ``lulc_array``.
+        Each value is the avoided pollutant load on that pixel in kg/yr,
+        or NODATA if any of the inputs have nodata on that pixel.
+    """
+    load_array = numpy.full(lulc_array.shape, NODATA, dtype=float)
+    valid_mask = (
+        (lulc_array != lulc_nodata) &
+        (retention_volume_array != NODATA))
+
+    # bin each value in the LULC array such that 
+    # lulc_array[i,j] == sorted_lucodes[lulc_index[i,j]]. thus, 
+    # emc_array[lulc_index[i,j]] is the EMC for the lucode at lulc_array[i,j]
+    lulc_index = numpy.digitize(lulc_array, sorted_lucodes, right=True)
+    # EMC for pollutant (mg/L) * 1000 (L/m^3) * 0.000001 (kg/mg) * 
+    # retention (m^3/yr) = pollutant load (kg/yr)
+    load_array[valid_mask] = (emc_array[lulc_index][valid_mask] * 
+        0.001 * retention_volume_array[valid_mask])
```

Comment:
You're making me re-think this, but my understanding was: by calling `digitize`, we get the _index_ of each lucode, and that makes it possible to do the same thing as a dictionary lookup, but with numpy. There unfortunately isn't a builtin numpy function to perform element-wise dictionary lookup. 
Instead, we sort the lucodes and make a corresponding EMC array such that `emc_array[i]` is the EMC value for the lucode at `sorted_lucodes[i]`. `emc_array[lulc_array]` wouldn't work because lucodes can be any integer, `10` and `20` are not indices in that array. So `numpy.digitize` converts the lucodes to their indices in `sorted_lucodes`. E.g. `[[10, 10], [20, 20]]` becomes `[[0, 0], [1, 1]]`. Then you can do `emc_array[lulc_index]` which is effectively doing a dictionary lookup but with arrays.
I got the idea from a similar thing [in NDR](https://github.com/natcap/invest/blob/36cc3d2e87a3ccdb934cf8fd36ba3a0fdeffd113/src/natcap/invest/ndr/ndr.py#L1022). Let me know if this still doesn't sound right!